File: packages/typescript/src/client/resources/CollectionsResource.ts
import type { components } from "../../generated/api";
import type { Page, PaginationOptions } from "../pagination/Page";
import { BaseResource } from "./BaseResource";

// Type aliases
type CollectionList = components["schemas"]["CollectionList"];
type CollectionDetail = components["schemas"]["CollectionDetail"];
type CollectionItem = components["schemas"]["CollectionItem"];
type CollectionType = components["schemas"]["CollectionType"];
type CollectionVisibility = components["schemas"]["CollectionVisibility"];
type CardCondition = components["schemas"]["CardCondition"];
type BulkAddItemsResult = components["schemas"]["BulkAddItemsResult"];
type BulkDeleteItemsResult = components["schemas"]["BulkDeleteItemsResult"];

// Response types
type CollectionListResponse = components["schemas"]["CollectionListResponse"];
type CollectionDetailResponse = components["schemas"]["CollectionDetailResponse"];
type CollectionCreateResponse = components["schemas"]["CollectionCreateResponse"];
type CollectionUpdateResponse = components["schemas"]["CollectionUpdateResponse"];
type CollectionDeleteResponse = components["schemas"]["CollectionDeleteResponse"];
type CollectionItemsListResponse = components["schemas"]["CollectionItemsListResponse"];
type CollectionItemResponse = components["schemas"]["CollectionItemResponse"];
type CollectionItemDeleteResponse = components["schemas"]["CollectionItemDeleteResponse"];
type BulkAddItemsResponse = components["schemas"]["BulkAddItemsResponse"];
type BulkDeleteItemsResponse = components["schemas"]["BulkDeleteItemsResponse"];

/** Options for creating a collection */
export interface CreateCollectionOptions {
  name: string;
  type?: CollectionType;
  visibility?: CollectionVisibility;
}

/** Options for updating a collection */
export interface UpdateCollectionOptions {
  name?: string;
  visibility?: CollectionVisibility;
}

/** Options for listing collection items */
export interface ListCollectionItemsOptions extends PaginationOptions {
  sortBy?: string;
  sortOrder?: string;
  search?: string;
}

/** Options for adding an item to a collection */
export interface AddCollectionItemOptions {
  cardId: string;
  quantity?: number;
  condition?: CardCondition;
  grade?: string;
  purchasePrice?: number;
  purchaseDate?: string;
  notes?: string;
}

/** Options for updating a collection item */
export interface UpdateCollectionItemOptions {
  quantity?: number;
  condition?: CardCondition;
  grade?: string;
  purchasePrice?: number;
  purchaseDate?: string;
  notes?: string;
}

/** Collections API - manage user card collections */
export class CollectionsResource extends BaseResource {
  /**
   * Get all collections for the authenticated user.
   *
   * @example
   * ```ts
   * const collections = await client.collections.list();
   * ```
   */
  async list(): Promise<CollectionList[]> {
    const response = await this.http.get<CollectionListResponse>("/Collections");
    return response.data ?? [];
  }

  /**
   * Get a single collection by ID.
   *
   * @param id - The collection GUID
   */
  async get(id: string): Promise<CollectionDetail> {
    const response = await this.http.get<CollectionDetailResponse>(
      `/Collections/${id}`
    );

    if (!response.data) {
      throw new Error("Collection not found");
    }

    return response.data;
  }

  /**
   * Create a new collection.
   *
   * @example
   * ```ts
   * const collection = await client.collections.create({
   *   name: 'My Rare Cards',
   *   type: 'Personal',
   *   visibility: 'Private'
   * });
   * ```
   */
  async create(options: CreateCollectionOptions): Promise<CollectionDetail> {
    const response = await this.http.post<CollectionCreateResponse>(
      "/Collections",
      {
        name: options.name,
        type: options.type,
        visibility: options.visibility,
      }
    );

    if (!response.data) {
      throw new Error("Failed to create collection");
    }

    return response.data;
  }

  /**
   * Update an existing collection.
   *
   * @param id - The collection GUID
   * @param options - Fields to update
   */
  async update(
    id: string,
    options: UpdateCollectionOptions
  ): Promise<CollectionDetail> {
    const response = await this.http.put<CollectionUpdateResponse>(
      `/Collections/${id}`,
      {
        name: options.name,
        visibility: options.visibility,
      }
    );

    if (!response.data) {
      throw new Error("Failed to update collection");
    }

    return response.data;
  }

  /**
   * Delete a collection.
   *
   * @param id - The collection GUID
   */
  async delete(id: string): Promise<void> {
    await this.http.delete<CollectionDeleteResponse>(`/Collections/${id}`);
  }

  /**
   * Get items in a collection with pagination.
   *
   * @param id - The collection GUID
   * @param options - Pagination and sorting options
   */
  async getItems(
    id: string,
    options: ListCollectionItemsOptions = {}
  ): Promise<Page<CollectionItem>> {
    const query = this.buildQuery({
      page: options.page,
      pageSize: options.pageSize,
      sortBy: options.sortBy,
      sortOrder: options.sortOrder,
      search: options.search,
    });

    const response = await this.http.get<CollectionItemsListResponse>(
      `/Collections/${id}/items`,
      query
    );

    return this.createPagedResponse(response, (paginationOpts) =>
      this.getItems(id, { ...options, ...paginationOpts })
    );
  }

  /**
   * Add an item to a collection.
   *
   * @param id - The collection GUID
   * @param options - Item details
   */
  async addItem(
    id: string,
    options: AddCollectionItemOptions
  ): Promise<CollectionItem> {
    const response = await this.http.post<CollectionItemResponse>(
      `/Collections/${id}/items`,
      {
        cardId: options.cardId,
        quantity: options.quantity,
        condition: options.condition,
        grade: options.grade,
        purchasePrice: options.purchasePrice,
        purchaseDate: options.purchaseDate,
        notes: options.notes,
      }
    );

    if (!response.data) {
      throw new Error("Failed to add item to collection");
    }

    return response.data;
  }

  /**
   * Update an item in a collection.
   *
   * @param collectionId - The collection GUID
   * @param itemId - The item GUID
   * @param options - Fields to update
   */
  async updateItem(
    collectionId: string,
    itemId: string,
    options: UpdateCollectionItemOptions
  ): Promise<CollectionItem> {
    const response = await this.http.put<CollectionItemResponse>(
      `/Collections/${collectionId}/items/${itemId}`,
      {
        quantity: options.quantity,
        condition: options.condition,
        grade: options.grade,
        purchasePrice: options.purchasePrice,
        purchaseDate: options.purchaseDate,
        notes: options.notes,
      }
    );

    if (!response.data) {
      throw new Error("Failed to update collection item");
    }

    return response.data;
  }

  /**
   * Remove an item from a collection.
   *
   * @param collectionId - The collection GUID
   * @param itemId - The item GUID
   */
  async removeItem(collectionId: string, itemId: string): Promise<void> {
    await this.http.delete<CollectionItemDeleteResponse>(
      `/Collections/${collectionId}/items/${itemId}`
    );
  }

  /**
   * Bulk add items to a collection.
   *
   * @param id - The collection GUID
   * @param items - Array of items to add
   */
  async bulkAddItems(
    id: string,
    items: AddCollectionItemOptions[]
  ): Promise<BulkAddItemsResult> {
    const response = await this.http.post<BulkAddItemsResponse>(
      `/Collections/${id}/items/bulk`,
      { items }
    );

    if (!response.data) {
      throw new Error("Failed to bulk add items");
    }

    return response.data;
  }

  /**
   * Bulk remove items from a collection.
   *
   * @param id - The collection GUID
   * @param itemIds - Array of item GUIDs to remove
   */
  async bulkRemoveItems(
    id: string,
    itemIds: string[]
  ): Promise<BulkDeleteItemsResult> {
    const response = await this.http.delete<BulkDeleteItemsResponse>(
      `/Collections/${id}/items/bulk`
    );

    // Note: The API uses DELETE with a body for bulk delete
    // We need to use the request method directly for this
    const result = await this.http.request<BulkDeleteItemsResponse>({
      method: "DELETE",
      path: `/Collections/${id}/items/bulk`,
      body: { itemIds },
    });

    if (!result.data) {
      throw new Error("Failed to bulk remove items");
    }

    return result.data;
  }
}

File: packages/csharp/src/PokeForge.SDK/Client/Resources/CollectionsResource.cs
namespace PokeForge.SDK.Client.Resources;

using System.Runtime.CompilerServices;
using PokeForge.SDK.Client.Http;
using PokeForge.SDK.Client.Pagination;
using PokeForge.SDK.Exceptions;
using PokeForge.SDK.Generated;

/// <summary>
/// Options for listing collection items.
/// </summary>
public sealed class ListCollectionItemsOptions : PaginationOptions
{
    /// <summary>
    /// Sort field.
    /// </summary>
    public string? SortBy { get; set; }

    /// <summary>
    /// Sort order (asc or desc).
    /// </summary>
    public string? SortOrder { get; set; }

    internal ListCollectionItemsOptions Clone() => new()
    {
        Page = Page,
        PageSize = PageSize,
        SortBy = SortBy,
        SortOrder = SortOrder
    };
}

/// <summary>
/// Collections API - manage user card collections.
/// Requires authentication.
/// </summary>
public sealed class CollectionsResource : BaseResource
{
    internal CollectionsResource(HttpClientWrapper http) : base(http)
    {
    }

    /// <summary>
    /// Get all collections for the authenticated user.
    /// Creates default collections (My Collection, Wishlist) on first access.
    /// </summary>
    public async Task<IReadOnlyList<CollectionList>> ListAsync(CancellationToken cancellationToken = default)
    {
        var response = await Http.GetAsync<CollectionListResponse>("/Collections", cancellationToken: cancellationToken)
            .ConfigureAwait(false);

        return response?.Data ?? Array.Empty<CollectionList>();
    }

    /// <summary>
    /// Get a single collection by ID.
    /// Returns collection if owned by user or if collection is public.
    /// </summary>
    public async Task<CollectionDetail> GetAsync(Guid id, CancellationToken cancellationToken = default)
    {
        var response = await Http.GetAsync<CollectionDetailResponse>($"/Collections/{id}", cancellationToken: cancellationToken)
            .ConfigureAwait(false);

        return response?.Data ?? throw new NotFoundException("Collection not found");
    }

    /// <summary>
    /// Create a new collection for the authenticated user.
    /// </summary>
    public async Task<CollectionDetail> CreateAsync(
        CreateCollectionRequest request,
        CancellationToken cancellationToken = default)
    {
        var response = await Http.PostAsync<CollectionCreateResponse>("/Collections", request, cancellationToken)
            .ConfigureAwait(false);

        return response?.Data ?? throw new PokeForgeException("Failed to create collection", 500);
    }

    /// <summary>
    /// Update an existing collection.
    /// Only the collection owner can update.
    /// </summary>
    public async Task<CollectionDetail> UpdateAsync(
        Guid id,
        UpdateCollectionRequest request,
        CancellationToken cancellationToken = default)
    {
        var response = await Http.PutAsync<CollectionUpdateResponse>($"/Collections/{id}", request, cancellationToken)
            .ConfigureAwait(false);

        return response?.Data ?? throw new PokeForgeException("Failed to update collection", 500);
    }

    /// <summary>
    /// Delete a collection (soft delete).
    /// Only the collection owner can delete. System collections and the default Wishlist cannot be deleted.
    /// </summary>
    public async Task DeleteAsync(Guid id, CancellationToken cancellationToken = default)
    {
        await Http.DeleteAsync($"/Collections/{id}", cancellationToken)
            .ConfigureAwait(false);
    }

    /// <summary>
    /// Get paginated items in a collection.
    /// </summary>
    public async Task<Page<CollectionItemSummary>> ListItemsAsync(
        Guid collectionId,
        ListCollectionItemsOptions? options = null,
        CancellationToken cancellationToken = default)
    {
        options ??= new ListCollectionItemsOptions();

        var query = BuildQuery(
            ("page", options.Page),
            ("pageSize", options.PageSize),
            ("sortBy", options.SortBy),
            ("sortOrder", options.SortOrder)
        );

        var response = await Http.GetAsync<CollectionItemsListResponse>($"/Collections/{collectionId}/items", query, cancellationToken)
            .ConfigureAwait(false);

        return CreatePagedResponse(
            response!,
            r => r.Data,
            r => r.Pagination,
            async (page, pageSize, ct) =>
            {
                var newOptions = options.Clone();
                newOptions.Page = page;
                newOptions.PageSize = pageSize;
                return await ListItemsAsync(collectionId, newOptions, ct).ConfigureAwait(false);
            });
    }

    /// <summary>
    /// Add a card to a collection.
    /// </summary>
    public async Task<CollectionItemSummary> AddItemAsync(
        Guid collectionId,
        AddCollectionItemRequest request,
        CancellationToken cancellationToken = default)
    {
        var response = await Http.PostAsync<CollectionItemResponse>($"/Collections/{collectionId}/items", request, cancellationToken)
            .ConfigureAwait(false);

        return response?.Data ?? throw new PokeForgeException("Failed to add item to collection", 500);
    }

    /// <summary>
    /// Update a collection item.
    /// </summary>
    public async Task<CollectionItemSummary> UpdateItemAsync(
        Guid collectionId,
        Guid itemId,
        UpdateCollectionItemRequest request,
        CancellationToken cancellationToken = default)
    {
        var response = await Http.PutAsync<CollectionItemResponse>($"/Collections/{collectionId}/items/{itemId}", request, cancellationToken)
            .ConfigureAwait(false);

        return response?.Data ?? throw new PokeForgeException("Failed to update collection item", 500);
    }

    /// <summary>
    /// Remove an item from a collection.
    /// </summary>
    public async Task RemoveItemAsync(
        Guid collectionId,
        Guid itemId,
        CancellationToken cancellationToken = default)
    {
        await Http.DeleteAsync($"/Collections/{collectionId}/items/{itemId}", cancellationToken)
            .ConfigureAwait(false);
    }

    /// <summary>
    /// Iterate through all items in a collection using IAsyncEnumerable.
    /// </summary>
    public async IAsyncEnumerable<CollectionItemSummary> ListAllItemsAsync(
        Guid collectionId,
        ListCollectionItemsOptions? options = null,
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        options ??= new ListCollectionItemsOptions();
        options.Page = 1;

        var page = await ListItemsAsync(collectionId, options, cancellationToken).ConfigureAwait(false);

        await foreach (var item in page.WithCancellation(cancellationToken))
        {
            yield return item;
        }
    }
}
